### PR TITLE
Add isNothingvalue proc

### PR DESCRIPTION
**Pull request by 12mohaned**
Resolve a simple Issue Mentioned here [17254](https://github.com/chapel-lang/chapel/issues/17254).

### CODE REVIEW

**Comment by mppf** [comment]
@12mohaned - thanks for looking at this. Actually we are hoping to have `isNothing`  added rather than `isNothingValue`. You might pick one of the others to look at - say - `isOwnedClass` - and then make a similar change adding `isNothing`. Note that at least one of the `isNothing` functions needs a documentation comment.

**Comment by mppf** [comment]
@12mohaned - right, now we have `isNothing(type t)` but we still need a version that takes in a value. You had that earlier - ` proc isNothingValue(e)   param return isNothingType(e.type); `. We need to have both.

Once you have both, you'll need to add a documentation comment before one of them and then add a test.

To learn more about documentation comments, see:
 * https://chapel-lang.org/docs/master/tools/chpldoc/chpldoc.html

To learn more about how to add a test, see:
 * https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst

**Comment by 12mohaned** [comment]
I added test files but could not use `start-test`(if it is possible to provide more info on it) command so tested them by compiling them normaly.

File: test/library/standard/Types/isNothing-proc.chpl [hunk 1]
```diff
@@ -0,0 +1,32 @@
+use Types;
+
+proc isNothing(type t)  param return isNothingType(t);
+proc isNothing(e)   param return isNothingType(e.type);
+
+
+proc myassert(param isNothing){
+writeln(isNothing);
```
Comment (mppf) [inline test/library/standard/Types/isNothing-proc.chpl:8]: let's have this function instead take in a `type t` argument and then you can print out both the type and the result of `isNothing` like this:

``` chapel
proc myassert(type t) {
  var dummyVar: t;
  writeln(t:string, " ", isNothing(t), " ", isNothing(dummyVar));
}
```

File: modules/standard/Types.chpl [hunk 2]
```diff
@@ -391,6 +393,14 @@ proc isConstAssignable(type t) param  return isConstAssignableType(t);
 pragma "no doc"
 proc isDefaultInitializable(type t) param return isDefaultInitializableType(t);
 
+/* Returns `true` if the argument is none or the `nothing` type.
+as defined by the language specification.*/
+pragma "no doc"
```
Comment (mppf) [inline modules/standard/Types.chpl:398]: Let's put none and nothing both in backquotes and remove the `pragma "no doc"`

**Comment by mppf** [comment]
> I added test files but could not use start-test(if it is possible to provide more info on it) command so tested them by compiling them normaly.

https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst#invoking-start-test should have the docs you need but try this:

```
make test-venv
./util/start_test test/library/standard/Types/isNothing-proc.chpl 
```

**Comment by 12mohaned** [comment]
> make test-venv
> ./util/start_test test/library/standard/Types/isNothing-proc.chpl

I did try the docs, whenever I call this make test-env, It gives the below error. 

make[2]: *** [-S] Error 1
make[1]: *** [/Users/mohanedmashaly/desktop/chapel-1.23.0/third-party/chpl-venv/install/darwin-x86_64/py2.7/chpl-virtualenv/chpl-test-reqs] Error 2

**Comment by mppf** [comment]
@12mohaned - right, you have Python 2.7 installed but it needs Python 3. You can install Python 3 with homebrew on a mac os X.

**Comment by 12mohaned** [comment]
@mppf I do have python 3.8 installed and did delete the python 2 path but it still gives me the same error.

**Comment by mppf** [comment]
It looks like you are using a release of Chapel rather than a build from a git clone? I think you are running into an issue we fixed after 1.23 was released. Can you try to do the `make` and `make test-venv` in a git clone of chapel?

**Comment by 12mohaned** [comment]
> It looks like you are using a release of Chapel rather than a build from a git clone? I think you are running into an issue we fixed after 1.23 was released. Can you try to do the `make` and `make test-venv` in a git clone of chapel?

It did work !, Thank you so much, I found a related problem in this [Issue](https://github.com/chapel-lang/chapel/issues/8766)

**Comment by mppf** [comment]
@12mohaned - great! So now have you been able to run `start_test` on your new test?

File: test/library/standard/Types/isNothing-proc.chpl [hunk 3]
```diff
@@ -0,0 +1,32 @@
+use Types;
+
+proc isNothing(type t)  param return isNothingType(t);
+proc isNothing(e)   param return isNothingType(e.type);
```
Comment (mppf) [inline test/library/standard/Types/isNothing-proc.chpl:4]: These functions should not be in the test since we want to test the ones in Types

**Comment by 12mohaned** [comment]
> @12mohaned - great! So now have you been able to run `start_test` on your new test?

Yes it worked partially but gave me the log gave me the following.

[Removing log file with duplicate name /Users/mohanedmashaly/desktop/chapel/test/Logs/mohanedmashaly.darwin.log]
[Starting Chapel regression tests - 210302.231833]
[git branch: 'isNothing_changes' (d01a6a2ded)]
[Warning: start_test not invoked from a subdirectory of $CHPL_HOME]
[starting directory: "/Users/mohanedmashaly/Desktop/chapel/util"]
[Logs directory: "/Users/mohanedmashaly/desktop/chapel/test/Logs"]
[log_file: "/Users/mohanedmashaly/desktop/chapel/test/Logs/mohanedmashaly.darwin.log"]
[CHPL_HOME: /Users/mohanedmashaly/desktop/chapel]
[host platform: darwin]
[target platform: darwin]
[valgrind: OFF]
[performance tests: OFF]
[compiler performance tests: OFF]
[number of trials: 1]
[performance graph generation: OFF]
[Error: Cannot find or execute compiler: /Users/mohanedmashaly/desktop/chapel/bin/darwin-x86_64/chpl]
[Done with tests - 210302.231833]
[Log file: /Users/mohanedmashaly/desktop/chapel/test/Logs/mohanedmashaly.darwin.log ]

[Test Summary - 210302.231833]
[Error: Cannot find or execute compiler: /Users/mohanedmashaly/desktop/chapel/bin/darwin-x86_64/chpl]
[Summary: #Successes = 0 | #Failures = 1 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]

**Comment by mppf** [comment]
Right, there the issue is I think `[Warning: start_test not invoked from a subdirectory of $CHPL_HOME]`. You should do `source util/setchplenv.bash` in your chapel checkout and then run `make` there and then try the `start_test` command again.

**Comment by 12mohaned** [comment]
It worked, Thanks again :=).

File: modules/standard/Types.chpl [hunk 4]
```diff
@@ -235,6 +235,8 @@ pragma "no doc"
 proc isPrimitiveValue(e) param  return isPrimitiveType(e.type);
 pragma "no doc"
 proc isEnumValue(e)      param  return isEnumType(e.type);
+
+
```
Comment (mppf) [inline modules/standard/Types.chpl:239]: Can you remove these extra spaces?

File: .vscode/settings.json [hunk 5]
```diff
@@ -4,5 +4,6 @@
     "chpl",
     "inited",
     "rtype"
-  ]
+  ],
+  "git.ignoreLimitWarning": true
```
Comment (mppf) [inline .vscode/settings.json:8]: I don't know how you ended up changing this file but I suspect it was accidental. Anyway, if we need to change this file, it should be in another PR. Can you remove the change here? (If necessary you can create a new PR and paste in your new code to make the history simpler).

Comment (12mohaned) [reply .vscode/settings.json:8]: No I did not mean to edit this file, I will remove it now.

**Comment by bradcray** [comment]
@12mohaned and @mppf : I'm marking this as post-release because even though it's low-impact and release, it feels more like a feature-freeze item, and not something that we'll suffer much from by not having it in the release.  That said, hopefully this can go in shortly after the release (?).

**Comment by bradcray** [comment]
@mppf: Pinging on this since you were running with it:  Is this good to go now or does more work remain?

File: test/library/standard/Types/isNothing-proc.chpl [hunk 6]
```diff
@@ -0,0 +1,23 @@
+use Types;
+
+proc myassert(type t){
+var Var: t;
```
Comment (mppf) [inline test/library/standard/Types/isNothing-proc.chpl:4]: let's indent the body of the function, please

File: test/library/standard/Types/isNothing-proc.chpl [hunk 7]
```diff
@@ -0,0 +1,23 @@
+use Types;
+
+proc myassert(type t){
```
Comment (mppf) [inline test/library/standard/Types/isNothing-proc.chpl:3]: please rename this function to `printIsNothing` since it doesn't call `assert`

**Comment by mppf** [comment]
@bradcray - it's very close - I left some minor feedback for @12mohaned to address and then it should be tested and merged

File: test/library/standard/Types/isNothing-proc.chpl (hunk 6 above)
Comment (mppf) [reply test/library/standard/Types/isNothing-proc.chpl:4]: (We use two spaces for each indent level)

**Comment by mppf** [comment]
hi @12mohaned - could you arrange for the PR to contain just one commit (since this is a really small change)? You could do that by rebasing (see https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/git.rst#how-to-modify-git-history - I tend to use `git rebase -i HEAD~<num>` myself with `<num>` replaced by the number of commits) or just by creating a new branch and PR with the change.
